File: dotnet/YapayGatewayLib/communication/RestV3.cs
using System;
using YapayGatewayLib.Model;
using System.Net.Http;
using System.Text;
using Newtonsoft.Json;

namespace YapayGatewayLib.Communication
{
    public class RestV3
    {

        public RestV3(string url)
        {
            this.url = url;
        }

        private string url;

        public string TransactionAuthorize(Credential credential, Transaction transaction)
        {

            var content = new StringContent(JsonConvert.SerializeObject(transaction), Encoding.UTF8, "application/json");

            var result = CreateHttpClient(credential).PostAsync(this.url + "/api/v3/transacao", content).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string TransactionQuery(Credential credential, string storeCode, long transactionNumber)
        {

            var result = CreateHttpClient(credential).GetAsync(this.url + "/api/v3/transacao/" + storeCode + "/" + transactionNumber).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string TransactionCapture(Credential credential, string storeCode, long transactionNumber, long value)
        {

            var result = CreateHttpClient(credential).PutAsync(this.url + "/api/v3/transacao/" + storeCode + "/" + transactionNumber, null).Result;


            return result.Content.ReadAsStringAsync().Result;
        }

        public string TransactionCancel(Credential credential, string storeCode, long transactionNumber, long value)
        {

            var result = CreateHttpClient(credential).PutAsync(this.url + "/api/v3/transacao/" + storeCode + "/" + transactionNumber, null).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string OneClickRegister(Credential credential, OneClickRegisterData registerData)
        {
            var content = new StringContent(JsonConvert.SerializeObject(registerData), Encoding.UTF8, "application/json");

            var result = CreateHttpClient(credential).PostAsync(this.url + "/api/v3/oneclick", content).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string OneClickQuery(Credential credential, string token)
        {
            var result = CreateHttpClient(credential).GetAsync(this.url + "/api/v3/oneclick/" + token).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string OneClickRegisterUpdate(Credential credential, string token, OneClickRegisterData registerData)
        {
            var content = new StringContent(JsonConvert.SerializeObject(registerData), Encoding.UTF8, "application/json");
            var result = CreateHttpClient(credential).PutAsync(this.url + "/api/v3/oneclick/" + token + "/alterar", content).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string OneClickAuthorize(Credential credential, string token, Transaction transaction)
        {
            var content = new StringContent(JsonConvert.SerializeObject(transaction), Encoding.UTF8, "application/json");

            var result = CreateHttpClient(credential).PostAsync(this.url + "/api/v3/oneclick/" + token + "/autorizar", content).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string RecurringPaymentRegister(Credential credential, RecurringPayment recurringPayment)
        {
            var content = new StringContent(JsonConvert.SerializeObject(recurringPayment), Encoding.UTF8, "application/json");

            var result = CreateHttpClient(credential).PostAsync(this.url + "/api/v3/recorrencia", content).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string RecurringPaymentQuery(Credential credential, string storeCode, long recurringPaymentNumber)
        {
            var result = CreateHttpClient(credential).GetAsync(this.url + "/api/v3/recorrencia/" + storeCode + "/" + recurringPaymentNumber).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        public string RecurringPaymentCancel(Credential credential, string storeCode, long recurringPaymentNumber)
        {

            var result = CreateHttpClient(credential).PutAsync(this.url + "/api/v3/recorrencia/" + storeCode + "/" + recurringPaymentNumber + "/cancelar", null).Result;

            return result.Content.ReadAsStringAsync().Result;
        }

        private HttpClient CreateHttpClient(Credential credential)
        {
            var byteArray = Encoding.ASCII.GetBytes(credential.user + ":" + credential.password);
            HttpClient client = new HttpClient();
            client.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Basic", Convert.ToBase64String(byteArray));
            return client;
        }
    }
}

File: ruby/yapay_gw/lib/yapay_gw/communication/rest_v3.rb
require 'http'
require_relative 'base_url'
require 'active_support'

class RestV3
    def transaction_authorize(credentials, transaction)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .post("#{BaseURL.get_url}/api/v3/transacao", :body => ActiveSupport::JSON.encode(transaction)) 
    end

    def transaction_query(credentials, transaction_number)
        transaction_query(credentials, credentials.store_code, transaction_number)
    end

    def transaction_query(credentials, store_code, transaction_number)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .get("#{BaseURL.get_url}/api/v3/transacao/#{store_code}/#{transaction_number}")
    end

    def transaction_capture(credentials, transaction_number, value)
        transaction_capture(credentials, credentials.store_code, transaction_number, value)
    end

    def transaction_capture(credentials, store_code, transaction_number, value)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .put("#{BaseURL.get_url}/api/v3/transacao/#{store_code}/#{transaction_number}/capturar")
    end

    def transaction_cancel(credentials, transaction_number, value)
        transaction_cancel(credentials, credentials.store_code, transaction_number, value)
    end

    def transaction_cancel(credentials, store_code, transaction_number, value)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .put("#{BaseURL.get_url}/api/v3/transacao/#{store_code}/#{transaction_number}/cancelar")
    end

    def oneclick_register(credentials, oneclick_register_data)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .post("#{BaseURL.get_url}/api/v3/oneclick", :body => ActiveSupport::JSON.encode(oneclick_register_data))
    end

    def oneclick_query(credentials, token)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .get("#{BaseURL.get_url}/api/v3/oneclick/#{token}")
    end

    def oneclick_register_update(credentials, token, oneclick_register_data)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .put("#{BaseURL.get_url}/api/v3/oneclick/#{token}/alterar", :body => ActiveSupport::JSON.encode(oneclick_register_data))
    end

    def oneclick_authorize(credentials, token, transaction)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .post("#{BaseURL.get_url}/api/v3/oneclick/#{token}/autorizar", :body => ActiveSupport::JSON.encode(oneclick_payment_data))
    end

    def recurring_payment_register(credentials, recurring_payment_data)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .post("#{BaseURL.get_url}/api/v3/recorrencia", :body => ActiveSupport::JSON.encode(recurring_payment_data))
    end

    def recurring_payment_query(credentials, recurring_payment_number)
        recurring_payment_query(credentials, credentials.store_code, recurring_payment_number);
    end

    def recurring_payment_query(credentials, store_code, recurring_payment_number)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .get("#{BaseURL.get_url}/api/v3/recorrencia/#{store_code}/#{recurring_payment_number}")
    end

    def recurring_payment_cancel(credentials, recurring_payment_number)
         recurring_payment_cancel(credentials, credentials.store_code, recurring_payment_number)
     end 

    def recurring_payment_cancel(credentials, store_code, recurring_payment_number)
        HTTP.basic_auth(:user => credentials.user, 
                        :pass => credentials.password)
            .headers(:user_agent => "Gateway Ruby Client", :content_type => "application/json")
            .put("#{BaseURL.get_url}/api/v3/recorrencia/#{store_code}/#{recurring_payment_number}/cancelar")
    end
end